Add testing framework

jasonrute opened the pull request:
I've added a testing framework for the trio server.  Basically I've created a mock lean process in Trio which runs according to a (linear) script.  I've included a test for the issue we already fixed where Lean stops waiting after we return a current_tasks response too early.

Also, to facilitate easier debugging, I've added another level of debugging called debug_bytes.  This prints the lean messages as they are sent over the channel.  Since `parse_response` in the trio server often crashes, this is a way to get at the message it is crashing on.  Also, it is useful for creating the mock lean scripts in the tests.

Curious to get your thoughts Patrick.

jasonrute commented:
Ok, prematurely PRed this, but it is all fixed now.

PatrickMassot commented:
Nice! I don't have time to read it carefully tonight (it's 11pm here), but the description of the mock Lean process sounds like what I intended to do at some point. The style of the test script reminds me of Behavior Driven Development. You could have a look at [pytest-bdd](https://github.com/pytest-dev/pytest-bdd) if you fancy that style.

PatrickMassot commented on test/test_trio_server/mock_lean.py:
```diff
@@ -0,0 +1,152 @@
+        self.stdout.put_data(message_bytes)
+
+    async def follow_script(self):
+        for step in self.script:
```
Why not defining a `run` method for each subclass of `LeanScriptStep` so that this loop body can be reduced to `step.run(self)`?

jasonrute replied:
How then would each LeanScriptStep relate to MockLeanProcess?  Would, say, LeanShouldGetRequest.run have access to the mock process?  This seems ultimately more complicated.  Here I'm thinking of LeanScriptStep as essentially an inductive data type and this is just a match statement over that data type.  (Of course lean doesn't have these concepts but that is the idea.)

jasonrute replied:
But I don't want to be too dogmatic either way, if you think that is best, we can do it.

PatrickMassot replied:
I indicated the relation in my message, by passing `self` to `step.run`. Now your answer exactly emphasize the issue, assuming you intended to write "python" instead of "lean" in the parenthesis. What you wrote is not python code. This is Lean (or Haskell, or Rust...) translated to python. I'll try to push a commit to this branch to show the difference.

jasonrute replied:
Yes, I meant to write Python.  Sorry, I didn't mean to imply what you said couldn't work  I agree it can.  I guess I have a resistance even in python, to classes depending on each other when it doesn't add much.  But I agree what I'm doing is already nonstandard with having a class for each step of the script.  Yes, how about you rewrite this as you think best and we can go from there.

PatrickMassot replied:
I tried this in 64ab19c. It does reinforce coupling between those classes, but they are very tightly related anyway. It seems to me that this solution will scale better.

PatrickMassot commented on test/test_trio_server/test_full_sync.py:
```diff
@@ -0,0 +1,47 @@
+        before continuing.
+        """
+        mock_lean_script = [
+            LeanShouldGetRequest({"file_name": "test.lean", "content": '--', "seq_num": 1, "command": "sync"}),
```
I think you'll be tired of writing all this Json. It would make more sense to either pass a subclass of `Request` object or have arguments to the `LeanShould...` constructor.

jasonrute replied:
In practice I usually just copy the json from the communication of a real world example.  

I also think this is a little dangerous though to always pass in the higher level object instead of the raw json.  Consider the following case:  We start working with, say, hole requests and discover that we forgot or misspelled a field in the hole request.  A test like this would ensure that exactly the right fields are sent in the correct way, instead of copying any errors we included in `.to_json()`.   Of course, this is a special case.  Usually, I'm just checking that Mock Lean and the interface code both get to the same point in their communication.  Maybe it is worth having two versions of LeanShould.  One which takes the JSON when it is needed, and another which takes the Request object.  

As for LeanSendsResponse, I think it is more important that we keep it as JSON at least for some of the cases.  For example, in later PRs I am testing that the interface doesn't crash when being send an infosource without a line or column.  Again, however, we could have multiple versions of LeanSendsRequest.  

What do you think?

PatrickMassot replied:
I think you are mixing tests here. There should be unit tests checking that `comands.py` does its job correctly, and then the tests we are currently looking at should trust it.

jasonrute replied:
Ok, I'll make an issue that we need to add unit tests for `commands.py`.  I guess there is no real reason to work on that until we complete https://github.com/leanprover-community/lean-client-python/issues/6.

jasonrute replied:
Ok, I added issue #11

jasonrute replied:
Ok, so I tried to refactor this so that `LeanShouldGetRequest` and `LeanShouldGetRequest` take objects from the `commands.py`.  It works for `LeanShouldGetRequest`.  Previously, the command was:
```python
LeanShouldGetRequest({"file_name": "test.lean", "content": '--', "seq_num": 1, "command": "sync"}),
```
and after it is
```python
LeanShouldGetRequest(SyncRequest(file_name="test.lean", content="--"), seq_num=1)
```
I'm not sure it is saving a lot of typing, but it does add some type safety.

As for `LeanShouldGetResponse`, that is more complicated.  I would need to write code that turns each `Response` object into a dictionary.  Maybe this can be revisited after #6 , but for now I'd like to just finish this up so I can fix the other bugs in the server.

PatrickMassot commented on test/test_trio_server/mock_lean.py:
```diff
@@ -0,0 +1,152 @@
+    """
+    Simulate taking a while to start up or processes a request.
+    """
+    seconds: float
```
If you use `@dataclass` with a single member and never subclass afterwards then I think you'd be better off using [NewType](https://mypy.readthedocs.io/en/stable/more_types.html#newtypes) (but below I'll suggest to add a method here, so this is a virtual comment).

jasonrute replied:
I'm not sure I see the advantage here of NewType in terms of performance (this code doesn't need to be performant) or readability, but sure I can switch them (assuming we don't go with the `.run()` method suggestion from above.)

jasonrute replied:
Is it possible to use NewType here since it is a subclass of LeanScriptStep?

PatrickMassot replied:
In that scenario it wouldn't be a subclass.

PatrickMassot commented on test/test_trio_server/mock_lean.py:
```diff
@@ -0,0 +1,152 @@
+                await self.assert_no_messages_received()
+
+
+async def start_with_mock_lean(lean_server: TrioLeanServer, script: List[LeanScriptStep]):
```
You could also monkeypatch `TrioLeanServer` in your test, but let's see if your way is convenient enough.

PatrickMassot commented on test/test_trio_server/test_full_sync.py:
```diff
@@ -0,0 +1,47 @@
+import trio.testing
+
+
+class TestFullSync(unittest.TestCase):
```
Oh! I strongly suggest we don't go back 15 years in time, and use pytest instead. Switching is very easy here. Just remove that line, and deindent the next function. Then remove the last two lines of this file. Sometimes progress is simply removing cruft.

jasonrute replied:
Sure.  I'm not so very familiar with different python testing frameworks.  I can use pytest.

jasonrute commented:
I switched to pytest.  I'll wait for your thoughts on my thoughts before making other changes here.

jasonrute commented:
Just so you know, I haven't forgot about this.  I got caught up in some other projects and personal stuff and I have to go to work today, but I'll get back to this soon.  Thanks for the changes!

PatrickMassot commented:
There is no hurry at all, I have plenty to do while waiting.

jasonrute commented:
Ok, I'm back at this.  (It only took me a month.)  I'll read through the code again and try to finish these PRs.

jasonrute commented:
Ok, added some more fixes, but I need to step merge (or rebase) master.  I tried it earlier and ran into some errors in the tests need to be fixed.

jasonrute commented:
Ok, I think we have addressed all the concerns.  @PatrickMassot can I get an approval on this?  (Also, note that I did a rebase, so if you are going to do anything with this branch, be aware.  I don't think you had any outstanding edits to this branch, so I thought it was safe.)

PatrickMassot commented on test/test_trio_server/mock_lean.py:
```diff
@@ -0,0 +1,175 @@
+        self.messages.extend(self.parse_message(m) for m in raw_messages)
+
+    async def assert_message_is_received(self, message_expected):
+        await trio.sleep(0.01)
```
Why do we need this explicit sleep (and the one below)?

jasonrute replied:
That is a great question.  I'm pretty sure it was necessary, but let me remember/investigate why and I'll add it as a comment to the code.

jasonrute replied:
Ok, I've addressed this in this commit (https://github.com/leanprover-community/lean-client-python/pull/3/commits/a585707f59fd7fcabf7a183ec0b3268dbf5df0ee).  In particular, 
1. I moved the `sleep` command into the `collect_stdin_messages` to avoid duplicated code.
2. I clearly explained what three goals the `sleep` command (in conjunction with `get_data_nowait()`) are intended to accomplish.
3. I made the waiting time a parameter which could be changed.  It might need to be extended if we are testing some complicated functionality which takes a while to compute.
4. I left the default of 0.01 seconds.  In my tests, 0.001 seconds also allowed the tests to pass but 0.0001 was too fast.  I thought 0.01 seconds was short enough to not make the tests take too long, but also gave an order of magnitude buffer to account for slow machines.

PatrickMassot replied:
I still don't understand. The whole point of this async thing is to avoid guessing how long we should wait for answers from Lean. Why do we need any sleep? Is this only for tests, simulating some delay in answers, or is this going to be needed in actual uses of this lean-python client?

jasonrute replied:
In short, it is only needed for tests because we don't want to wait forever for the Lean Python client to return information which it may never do.

Remember, this is "Mock Lean", so I'm simulating a fake Lean processes waiting for our Python client to set information.  A real Lean client would make a blocking call to read the data and wait as long as necessary.  That could be accomplished via `self.stdin.get_data()` in `collect_stdin_messages`, but the problem is that I don't know if there are any messages and I could be left waiting forever.  This is used to support `LeanShouldGetRequest` and `LeanShouldNotGetRequest`.  Those test that either Lean receives a request (and it a particular form) or that it doesn't receive a request, so I need the ability to stop waiting.

So what I'm doing here is waiting for our Python client for a little bit (enough time for our Python client to run and send a message if it is going to), and then checking if the message is sent in a non-blocking manner.

Another way I could possibly implement this is via a blocking call, `self.stdin.get_data()`, and then use a feature of Trio testing framework that has some way to stop execution of a blocked coroutine after a certain time.  I think this might exist, however, I'd have to look into that more.  It is not clear to me how to implement this.

jasonrute replied:
So like I said, an alternative way to do this would be with [cancellation and timeouts](https://trio.readthedocs.io/en/stable/reference-core.html#cancellation-and-timeouts).  I would still need to look into how to do this more.  Would you prefer that approach?

jasonrute replied:
I could do something like this inside `collect_stdin_messages`:
```
with trio.move_on_after(wait_time_seconds):
    try:
        data = await self.stdin.get_data()
    except Cancelled:
        return None
   
    self.partial_message = raw_messages.pop()
    self.messages.extend(self.parse_message(m) for m in raw_messages)
```

jasonrute replied:
Ok, I just pushed [a commit using the above snippet](https://github.com/leanprover-community/lean-client-python/pull/3/commits/1c9caac4dd845324d2e2a208a8c69dd3f5121d87).  I think it will clear up a lot of the confusion.

PatrickMassot replied:
Yes, I think this expresses what's going on much more directly.
